File: Flutter/Merchant/Helper Codes/objc/Minkasu2FAUtils.m
#import "Minkasu2FAUtils.h"
#import "Minkasu2FA/Minkasu2FACallbackDelegate.h"
#import <Minkasu2FA/Minkasu2FAHeader.h>
@import webview_flutter_wkwebview;

static NSString *channelName = @"minkasu2fa_method_channel";

@interface Minkasu2FAUtils () <Minkasu2FACallbackDelegate>
@property(nonatomic, strong) FlutterMethodChannel *channel;
@end

@implementation Minkasu2FAUtils

+ (instancetype)sharedInstance {
    static Minkasu2FAUtils *sharedInstance = nil;
    static dispatch_once_t onceToken; // onceToken = 0
    dispatch_once(&onceToken, ^{
        sharedInstance = [[super alloc] init];
    });
    
    return sharedInstance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
    }
    return self;
}

- (void)setUpMinkasu2FAWithController:(FlutterViewController *)controller {
    self.channel =
    [FlutterMethodChannel methodChannelWithName:channelName
                                binaryMessenger:controller.binaryMessenger];
    [self handleMethodCalls:nil];
}

- (void)setUpMinkasu2FAWithEngine:(FlutterEngine *)engine {
    self.channel =
    [FlutterMethodChannel methodChannelWithName:channelName
                                binaryMessenger:engine.binaryMessenger];
    [self handleMethodCalls:engine];
}

- (void)handleMethodCalls:(nullable FlutterEngine *)engine {
    [self.channel setMethodCallHandler:^(FlutterMethodCall *call,
                                         FlutterResult result) {
        if ([@"initMinkasu2FASDK" isEqualToString:call.method]) {
            NSDictionary *arguments = call.arguments;
            NSNumber *number = arguments[@"webViewId"];
            NSDictionary *configDictionary = arguments[@"config"];
            
            if ((arguments && arguments != nil) && (number && number != nil) &&
                (configDictionary && configDictionary != nil)) {
                Minkasu2FAConfig *config = [[Minkasu2FAUtils sharedInstance]
                                            constructConfigFrom:configDictionary];
                WKWebView *webView = nil;
                if (engine && engine != nil) {
                    webView = [[Minkasu2FAUtils sharedInstance]
                               fetchWebViewWithId:[number longValue]
                               andEngine:engine];
                } else {
                    webView = [[Minkasu2FAUtils sharedInstance]
                               fetchWebViewWithId:[number longValue]];
                }
                if (webView && webView != nil &&
                    ![webView isKindOfClass:[NSNull class]]) {
                    [[Minkasu2FAUtils sharedInstance] initSDKwithWebView:webView
                                                                  config:config
                                                               andResult:result];
                } else {
                    result(@{
                        @"status" : @"failed",
                        @"code" : @"MK2FA_F101",
                        @"message" : @"Error in accessing webview"
                    });
                }
            } else {
                result(@{
                    @"status" : @"failed",
                    @"code" : @"MK2FA_F101",
                    @"message" : @"Invalid arguments"
                });
            }
        } else {
            result(FlutterMethodNotImplemented);
        }
    }];
}

- (WKWebView *)fetchWebViewWithId:(long)webViewId
                        andEngine:(FlutterEngine *)engine {
    WKWebView *webView = (WKWebView *)[FWFWebViewFlutterWKWebViewExternalAPI
                                       webViewForIdentifier:webViewId
                                       withPluginRegistry:engine];
    return webView;
}

- (WKWebView *)fetchWebViewWithId:(long)webViewId {
    FlutterAppDelegate *flutterAppDelegate =
    (FlutterAppDelegate *)[UIApplication sharedApplication].delegate;
    if (!flutterAppDelegate || flutterAppDelegate == nil ||
        [flutterAppDelegate isKindOfClass:[NSNull class]]) {
        return nil;
    }
    WKWebView *webView = (WKWebView *)[FWFWebViewFlutterWKWebViewExternalAPI
                                       webViewForIdentifier:webViewId
                                       withPluginRegistry:flutterAppDelegate];
    return webView;
}

- (void)initSDKwithWebView:(WKWebView *)webView
                    config:(Minkasu2FAConfig *)config
                 andResult:(FlutterResult)result {
    NSError *error = nil;
    NSString *hybridSDKDetails = @"{\"platform\": \"flutter\"}";
    [Minkasu2FA initHybridSDKWithWKWebView:webView andConfiguration:config hybridSDKDetails: hybridSDKDetails inViewController:nil error:&error];
    if (error) {
        result(@{
            @"status" : @"failed",
            @"code" : error.domain,
            @"message" : error.userInfo[@"NSLocalizedDescription"],
        });
        return;
    }
    
    result(@{
        @"status" : @"success",
    });
}

- (Minkasu2FAConfig *)constructConfigFrom:(NSDictionary *)configDictionary {
    NSString *configId = configDictionary[@"id"];
    NSString *merchantCustomerId = configDictionary[@"merchantCustomerId"];
    NSString *token = configDictionary[@"token"];
    
    NSDictionary *customerDictionary = configDictionary[@"customerInfo"];
    Minkasu2FACustomerInfo *customerInfo =
    [self constructCustomerInfoFrom:customerDictionary];
    
    NSDictionary *orderInfoDictionary = configDictionary[@"orderInfo"];
    Minkasu2FAOrderInfo *orderInfo =
    [self constructOrderInfoFrom:orderInfoDictionary];
    
    NSNumber *sdkModeValue = configDictionary[@"sdkMode"];
    Minkasu2FASDKMode sdkMode = [self constructSDKModeFrom:sdkModeValue];
    
    NSDictionary *customThemeDictionary = configDictionary[@"customTheme"];
    Minkasu2FACustomTheme *customTheme = nil;
    if (customThemeDictionary && customThemeDictionary != nil &&
        ![customThemeDictionary isKindOfClass:[NSNull class]]) {
        customTheme = [self constructCustomThemeFrom:customThemeDictionary];
    }
    
    Minkasu2FAConfig *config = [Minkasu2FAConfig new];
    config._id = configId;
    config.token = token;
    config.merchantCustomerId = merchantCustomerId;
    config.customerInfo = customerInfo;
    config.sdkMode = sdkMode;
    config.orderInfo = orderInfo;
    config.customTheme = customTheme;
    config.delegate = self;
    
    return config;
}

- (Minkasu2FACustomerInfo *)constructCustomerInfoFrom:
(NSDictionary *)customerInfoDictionary {
    NSString *firstName = customerInfoDictionary[@"firstName"];
    NSString *lastName = customerInfoDictionary[@"lastName"];
    NSString *middleName = customerInfoDictionary[@"middleName"];
    NSString *email = customerInfoDictionary[@"email"];
    NSString *phone = customerInfoDictionary[@"phone"];
    
    NSDictionary *addressDictionary = customerInfoDictionary[@"address"];
    Minkasu2FAAddress *address = nil;
    
    if (addressDictionary && addressDictionary != nil &&
        ![addressDictionary isKindOfClass:[NSNull class]]) {
        address = [self constructAddressFrom:addressDictionary];
    }
    
    Minkasu2FACustomerInfo *customerInfo = [Minkasu2FACustomerInfo new];
    customerInfo.firstName = firstName;
    customerInfo.lastName = lastName;
    if (middleName && middleName != nil &&
        ![middleName isKindOfClass:[NSNull class]]) {
        customerInfo.middleName = middleName;
    }
    customerInfo.email = email;
    customerInfo.phone = phone;
    customerInfo.address = address;
    
    return customerInfo;
}

- (Minkasu2FAAddress *)constructAddressFrom:(NSDictionary *)addressDictionary {
    NSString *line1 = addressDictionary[@"line1"];
    NSString *line2 = addressDictionary[@"line2"];
    NSString *line3 = addressDictionary[@"line3"];
    NSString *city = addressDictionary[@"city"];
    NSString *state = addressDictionary[@"state"];
    NSString *country = addressDictionary[@"country"];
    NSString *zipCode = addressDictionary[@"zipCode"];
    
    Minkasu2FAAddress *address = [Minkasu2FAAddress new];
    if (line1 && line1 != nil && ![line1 isKindOfClass:[NSNull class]]) {
        address.line1 = line1;
    }
    if (line2 && line2 != nil && ![line2 isKindOfClass:[NSNull class]]) {
        address.line3 = line2;
    }
    if (line3 && line3 != nil && ![line3 isKindOfClass:[NSNull class]]) {
        address.line3 = line3;
    }
    if (city && city != nil && ![city isKindOfClass:[NSNull class]]) {
        address.city = city;
    }
    if (state && state != nil && ![state isKindOfClass:[NSNull class]]) {
        address.state = state;
    }
    if (country && country != nil && ![country isKindOfClass:[NSNull class]]) {
        address.country = country;
    }
    if (zipCode && zipCode != nil && ![zipCode isKindOfClass:[NSNull class]]) {
        address.zipCode = zipCode;
    }
    
    return address;
}

- (Minkasu2FAOrderInfo *)constructOrderInfoFrom:
(NSDictionary *)orderInfoDictionary {
    NSString *orderId = orderInfoDictionary[@"orderId"];
    NSString *orderDetails = orderInfoDictionary[@"orderDetails"];
    NSString *billingCategory = orderInfoDictionary[@"billingCategory"];
    
    Minkasu2FAOrderInfo *orderInfo = [Minkasu2FAOrderInfo new];
    orderInfo.orderId = orderId;
    
    if (orderDetails && orderDetails != nil &&
        ![orderDetails isKindOfClass:[NSNull class]]) {
        orderInfo.orderDetails = orderDetails;
    }
    if (billingCategory && billingCategory != nil &&
        ![billingCategory isKindOfClass:[NSNull class]]) {
        orderInfo.billingCategory = billingCategory;
    }
    
    return orderInfo;
}

- (Minkasu2FASDKMode)constructSDKModeFrom:(NSNumber *)sdkModeValue {
    if ([sdkModeValue intValue] == 0) {
        return MINKASU2FA_PRODUCTION_MODE;
    }
    return MINKASU2FA_SANDBOX_MODE;
}

- (Minkasu2FACustomTheme *)constructCustomThemeFrom:
(NSDictionary *)customThemeDictionary {
    NSNumber *navigationBarColor = customThemeDictionary[@"navigationBarColor"];
    NSNumber *navigationBarTextColor =
    customThemeDictionary[@"navigationBarTextColor"];
    NSNumber *darkModeNavigationBarColor =
    customThemeDictionary[@"darkModeNavigationBarColor"];
    NSNumber *darkModeNavigationBarTextColor =
    customThemeDictionary[@"darkModeNavigationBarTextColor"];
    BOOL supportDarkMode = customThemeDictionary[@"supportDarkMode"];
    
    Minkasu2FACustomTheme *customTheme = [Minkasu2FACustomTheme new];
    customTheme.navigationBarColor = [self
                                      getARGBComponentsFromColor:[navigationBarColor unsignedIntegerValue]];
    customTheme.navigationBarTextColor =
    [self getARGBComponentsFromColor:[navigationBarTextColor
                                      unsignedIntegerValue]];
    customTheme.darkModeNavigationBarColor =
    [self getARGBComponentsFromColor:[darkModeNavigationBarColor
                                      unsignedIntegerValue]];
    customTheme.darkModeNavigationBarTextColor =
    [self getARGBComponentsFromColor:[darkModeNavigationBarTextColor
                                      unsignedIntegerValue]];
    customTheme.supportDarkMode = supportDarkMode;
    
    return customTheme;
}

- (NSDictionary *)convertCallBackInfoToDictionaryFrom:
(Minkasu2FACallbackInfo *)callBackInfo {
    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
    dictionary[@"infoType"] =
    [NSNumber numberWithInteger:callBackInfo.infoType];
    dictionary[@"data"] = callBackInfo.data;
    
    return dictionary;
}

- (UIColor *)getARGBComponentsFromColor:(NSUInteger)colorValue {
    
    CGFloat alpha = ((colorValue >> 24) & 0xFF) / 255.0;
    CGFloat red = ((colorValue >> 16) & 0xFF) / 255.0;
    CGFloat green = ((colorValue >> 8) & 0xFF) / 255.0;
    CGFloat blue = (colorValue & 0xFF) / 255.0;
    
    UIColor *color = [UIColor colorWithRed:red
                                     green:green
                                      blue:blue
                                     alpha:alpha];
    return color;
}

#pragma mark - Minkasu2FACallback Delegate

- (void)minkasu2FACallback:(Minkasu2FACallbackInfo *)minkasu2FACallbackInfo {
    NSDictionary *callBackDictionary =
    [self convertCallBackInfoToDictionaryFrom:minkasu2FACallbackInfo];
    [self.channel invokeMethod:@"receiveMinkasu2FACallbackInfo"
                     arguments:callBackDictionary];
}

@end

File: Flutter/Partner/Helper Codes/swift/Minkasu2FAUtils.swift
import Flutter
import Minkasu2FA
import webview_flutter_wkwebview
import WebKit

class Minkasu2FAUtils:NSObject, Minkasu2FACallbackDelegate {
    static let channelName = "minkasu2fa_method_channel"
    static let shared = Minkasu2FAUtils()
    
    var channel: FlutterMethodChannel?
    
    override private init() {
        super.init()
    }
    
    func setUpMinkasu2FAWithEngine(engine: FlutterEngine) {
        self.channel = FlutterMethodChannel(name: Minkasu2FAUtils.channelName, binaryMessenger: engine.binaryMessenger)
        handleMethodCalls(flutterEngine: engine)
    }
    
    func setUpMinkasu2FAWithController(controller: FlutterViewController) {
        self.channel = FlutterMethodChannel(name: Minkasu2FAUtils.channelName, binaryMessenger: controller.binaryMessenger)
        handleMethodCalls()
    }
    
    private func handleMethodCalls(flutterEngine: FlutterEngine? = nil) {
        self.channel?.setMethodCallHandler { (call, result) in
            if call.method == "initMinkasu2FASDK" {
                guard let arguments = call.arguments as? [String: Any],
                      let configDictionary = arguments["config"] as? [String: Any],
                      let webViewIdNum = arguments["webViewId"] as? NSNumber
                else {
                    result([
                        "status": "failed",
                        "code": "MK2FA_F101",
                        "message": "Invalid arguments"
                    ])
                    return
                }
                let webViewId = Int64(truncating: webViewIdNum)
                let webView: WKWebView?
                
                if let engine = flutterEngine {
                    webView = Minkasu2FAUtils.shared.fetchWebViewWithId(webViewId: webViewId, engine: engine)
                } else {
                    webView = Minkasu2FAUtils.shared.fetchWebViewWithId(webViewId: webViewId)
                }
                
                guard let unwrappedWebView = webView else {
                    result([
                        "status": "failed",
                        "code": "MK2FA_F101",
                        "message": "Error in accessing webview"
                    ])
                    return
                }
                
                let config = Minkasu2FAUtils.shared.constructConfig(from: configDictionary)
                Minkasu2FAUtils.shared.initializeMinkasu2FASDK(webView: unwrappedWebView, config: config, result: result)
                
            } else {
                result(FlutterMethodNotImplemented)
            }
        }
    }
    
    private func fetchWebViewWithId(webViewId: Int64) -> WKWebView? {
        let flutterAppDelegate2 = UIApplication.shared.delegate as? FlutterPluginRegistry
        guard let flutterAppDelegate = UIApplication.shared.delegate as? FlutterPluginRegistry,
              let webView = FWFWebViewFlutterWKWebViewExternalAPI.webView(forIdentifier: webViewId, withPluginRegistry: flutterAppDelegate) else {
            return nil
        }
        return webView
    }
    
    private func fetchWebViewWithId(webViewId: Int64, engine: FlutterEngine) -> WKWebView? {
        let webView = FWFWebViewFlutterWKWebViewExternalAPI.webView(forIdentifier: webViewId, withPluginRegistry: engine)
        return webView
    }
    
    private func initializeMinkasu2FASDK(webView: WKWebView, config: Minkasu2FAConfig, result: FlutterResult) {
        do {
            try try Minkasu2FA.initWith(webView, andConfiguration: config)
            result([
                "status": "success"
            ])
        } catch let error as NSError {
            result([
                "status": "failed",
                "code": error.domain,
                "message": error.localizedDescription
            ])
        }
    }
    
    private func constructConfig(from configDict: [String: Any]) -> Minkasu2FAConfig {
        let configId = configDict["id"] as? String
        let merchantCustomerId = configDict["merchantCustomerId"] as? String
        let token = configDict["token"] as? String
        
        let customerDict = configDict["customerInfo"] as? [String: Any]
        let customerInfo = constructCustomerInfo(from: customerDict)
        
        let orderDict = configDict["orderInfo"] as? [String: Any]
        let orderInfo = constructOrderInfo(from: orderDict)
        
        let partnerInfoDict = configDict["partnerInfo"] as? [String: Any]
        let partnerInfo = constructPartnerInfo(from: partnerInfoDict)
        
        let sdkModeValue = configDict["sdkMode"] as? Int
        let sdkMode = constructSDKMode(from: sdkModeValue)
        
        let customThemeDict = configDict["customTheme"] as? [String: Any]
        let customTheme = customThemeDict != nil ? constructCustomTheme(from: customThemeDict!) : nil
        
        let config = Minkasu2FAConfig()
        config._id = configId
        config.token = token
        config.merchantCustomerId = merchantCustomerId
        config.customerInfo = customerInfo
        config.sdkMode = sdkMode
        config.orderInfo = orderInfo
        config.customTheme = customTheme
        config.partnerInfo = partnerInfo
        config.delegate = self
        
        return config
    }
    
    private func constructCustomerInfo(from dict: [String: Any]?) -> Minkasu2FACustomerInfo? {
        guard let dict = dict else { return nil }
        
        let firstName = dict["firstName"] as? String
        let lastName = dict["lastName"] as? String
        let email = dict["email"] as? String
        let phone = dict["phone"] as? String
        
        var address: Minkasu2FAAddress? = nil
        if let addressDict = dict["address"] as? [String: Any] {
            address = constructAddress(from: addressDict)
        }
        
        let customerInfo = Minkasu2FACustomerInfo()
        customerInfo.firstName = firstName
        customerInfo.lastName = lastName
        customerInfo.email = email
        customerInfo.phone = phone
        customerInfo.address = address
        
        return customerInfo
    }
    
    private func constructAddress(from dict: [String: Any]) -> Minkasu2FAAddress {
        let address = Minkasu2FAAddress()
        
        if let line1 = dict["line1"] as? String {
            address.line1 = line1
        }
        if let line2 = dict["line2"] as? String {
            address.line2 = line2
        }
        if let city = dict["city"] as? String {
            address.city = city
        }
        if let state = dict["state"] as? String {
            address.state = state
        }
        if let country = dict["country"] as? String {
            address.country = country
        }
        if let zipCode = dict["zipCode"] as? String {
            address.zipCode = zipCode
        }
        
        return address
    }
    
    private func constructOrderInfo(from dict: [String: Any]?) -> Minkasu2FAOrderInfo? {
        guard let dict = dict else { return nil }
        
        let orderInfo = Minkasu2FAOrderInfo()
        orderInfo.orderId = dict["orderId"] as? String
        orderInfo.orderDetails = dict["orderDetails"] as? String
        orderInfo.billingCategory = dict["billingCategory"] as? String
        
        return orderInfo
    }
    
    private func constructPartnerInfo(from dict: [String: Any]?) -> Minkasu2FAPartnerInfo? {
        guard let dict = dict else { return nil }
        
        let partnerInfo = Minkasu2FAPartnerInfo()
        if let merchantId = dict["merchantId"] as? String {
            partnerInfo.merchantId = merchantId
        }
        if let merchantName = dict["merchantName"] as? String {
            partnerInfo.merchantName = merchantName
        }
        if let transactionId = dict["transactionId"] as? String {
            partnerInfo.transactionId = transactionId
        }
        
        return partnerInfo
    }
    
    private func constructSDKMode(from value: Int?) -> Minkasu2FASDKMode {
        guard let value = value else { return Minkasu2FASDKMode.MINKASU2FA_SANDBOX_MODE }
        return value == 0 ? Minkasu2FASDKMode.MINKASU2FA_PRODUCTION_MODE : Minkasu2FASDKMode.MINKASU2FA_SANDBOX_MODE
    }
    
    private func constructCustomTheme(from dict: [String: Any]) -> Minkasu2FACustomTheme {
        let theme = Minkasu2FACustomTheme()
        
        if let navBarColorValue = dict["navigationBarColor"] as? UInt {
            theme.navigationBarColor = getARGBColorComponents(from: navBarColorValue)
        }
        if let navBarTextColorValue = dict["navigationBarTextColor"] as? UInt {
            theme.navigationBarTextColor = getARGBColorComponents(from: navBarTextColorValue)
        }
        if let darkNavBarColorValue = dict["darkModeNavigationBarColor"] as? UInt {
            theme.darkModeNavigationBarColor = getARGBColorComponents(from: darkNavBarColorValue)
        }
        if let darkNavBarTextColorValue = dict["darkModeNavigationBarTextColor"] as? UInt {
            theme.darkModeNavigationBarTextColor = getARGBColorComponents(from: darkNavBarTextColorValue)
        }
        if let supportDarkMode = dict["supportDarkMode"] as? Bool {
            theme.supportDarkMode = supportDarkMode
        }
        
        return theme
    }
    
    private func getARGBColorComponents(from colorValue: UInt) -> UIColor {
        let alpha = CGFloat((colorValue >> 24) & 0xFF) / 255.0
        let red = CGFloat((colorValue >> 16) & 0xFF) / 255.0
        let green = CGFloat((colorValue >> 8) & 0xFF) / 255.0
        let blue = CGFloat(colorValue & 0xFF) / 255.0
        
        return UIColor(red: red, green: green, blue: blue, alpha: alpha)
    }
    
    private func convertCallbackInfoToDictionary(from callbackInfo: Minkasu2FACallbackInfo) -> [String: Any] {
        let dictionary: [String: Any] = [
            "infoType": callbackInfo.infoType,
            "data": callbackInfo.data
        ]
        
        return dictionary
    }
    
    func minkasu2FACallback(_ minkasu2FACallbackInfo: Minkasu2FACallbackInfo) {
        let callbackDictionary = convertCallbackInfoToDictionary(from: minkasu2FACallbackInfo)
        self.channel?.invokeMethod("receiveMinkasu2FACallbackInfo", arguments: callbackDictionary)
    }
}
